include/cuco/detail/static_set/kernels.cuh

```diff
@@ -159,24 +106,42 @@ __global__ void insert(InputIterator first,
  * @tparam BlockSize Number of threads in each block
  * @tparam InputIterator Device accessible input iterator whose `value_type` is
  * convertible to the `value_type` of the data structure
+ * @tparam StencilIt Device accessible random access iterator whose value_type is
+ * convertible to Predicate's argument type
+ * @tparam Predicate Unary predicate callable whose return type must be convertible to `bool`
+ * and argument type is convertible from `std::iterator_traits<StencilIt>::value_type`
  * @tparam Ref Type of non-owning device ref allowing access to storage
  *
  * @param first Beginning of the sequence of input elements
  * @param n Number of input elements
+ * @param stencil Beginning of the stencil sequence
+ * @param pred Predicate to test on every element in the range `[s, s + n)`
  * @param ref Non-owning set device ref used to access the slot storage
  */
-template <int32_t CGSize, int32_t BlockSize, typename InputIterator, typename Ref>
-__global__ void insert_async(InputIterator first, cuco::detail::index_type n, Ref ref)
+template <int32_t CGSize,
+          int32_t BlockSize,
+          typename InputIterator,
+          typename StencilIt,
+          typename Predicate,
+          typename Ref>
+__global__ void insert_if_n(
+  InputIterator first, cuco::detail::index_type n, StencilIt stencil, Predicate pred, Ref ref)
 {
-  namespace cg = cooperative_groups;
-
-  auto tile                                  = cg::tiled_partition<CGSize>(cg::this_thread_block());
   cuco::detail::index_type const loop_stride = gridDim.x * BlockSize / CGSize;
   cuco::detail::index_type idx               = (BlockSize * blockIdx.x + threadIdx.x) / CGSize;
 
+  [[maybe_unused]] auto const tile =
```

Comment:
Instead of `[[maybe_unsed]]`, can't we just put the construction of `tile` in the `else` branch?

Reply:
> put the construction of tile in the else branch

I thought about this, and my concern at that point was that the tiles would be constructed in each loop iteration instead of only once before the loop. Just benchmarked the impact of this change and the difference is indeed negligible. 

# static_set_insert_uniform_multiplicity

## [0] Quadro RTX 8000

|  Key  |  Distribution  |  Multiplicity  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|-------|----------------|----------------|------------|-------------|------------|-------------|--------------|---------|----------|
|  I32  |    UNIFORM     |       1        |  68.832 ms |       1.25% |  69.022 ms |       1.96% |   189.879 us |   0.28% |   PASS   |
|  I32  |    UNIFORM     |       2        |  49.334 ms |       7.23% |  48.275 ms |       0.06% | -1058.434 us |  -2.15% |   FAIL   |
|  I32  |    UNIFORM     |       4        |  40.042 ms |       0.06% |  40.057 ms |       0.06% |    14.711 us |   0.04% |   PASS   |
|  I32  |    UNIFORM     |       8        |  36.544 ms |       0.07% |  36.597 ms |       0.32% |    52.684 us |   0.14% |   FAIL   |
|  I32  |    UNIFORM     |       16       |  34.085 ms |       0.04% |  34.078 ms |       0.06% |    -6.754 us |  -0.02% |   PASS   |
|  I64  |    UNIFORM     |       1        |  70.714 ms |       0.06% |  70.648 ms |       0.05% |   -66.304 us |  -0.09% |   FAIL   |
|  I64  |    UNIFORM     |       2        |  50.163 ms |       0.11% |  50.180 ms |       0.14% |    17.108 us |   0.03% |   PASS   |
|  I64  |    UNIFORM     |       4        |  41.962 ms |       0.14% |  41.961 ms |       0.16% |    -1.531 us |  -0.00% |   PASS   |
|  I64  |    UNIFORM     |       8        |  38.209 ms |       0.13% |  38.192 ms |       0.08% |   -17.092 us |  -0.04% |   PASS   |
|  I64  |    UNIFORM     |       16       |  36.917 ms |       0.41% |  36.919 ms |       0.27% |     1.535 us |   0.00% |   PASS   |

# static_set_insert_unique_occupancy

## [0] Quadro RTX 8000

|  Key  |  Distribution  |  Occupancy  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |        Diff |   %Diff |  Status  |
|-------|----------------|-------------|------------|-------------|------------|-------------|-------------|---------|----------|
|  I32  |     UNIQUE     |     0.1     |  79.365 ms |       0.11% |  79.325 ms |       0.07% |  -39.857 us |  -0.05% |   PASS   |
|  I32  |     UNIQUE     |     0.2     |  76.472 ms |       0.05% |  76.389 ms |       0.06% |  -83.348 us |  -0.11% |   FAIL   |
|  I32  |     UNIQUE     |     0.3     |  75.482 ms |       0.06% |  75.445 ms |       0.04% |  -36.994 us |  -0.05% |   FAIL   |
|  I32  |     UNIQUE     |     0.4     |  75.065 ms |       0.04% |  75.015 ms |       0.03% |  -50.085 us |  -0.07% |   FAIL   |
|  I32  |     UNIQUE     |     0.5     |  75.009 ms |       0.04% |  74.976 ms |       0.04% |  -32.227 us |  -0.04% |   FAIL   |
|  I32  |     UNIQUE     |     0.6     |  75.264 ms |       0.04% |  75.212 ms |       0.05% |  -52.269 us |  -0.07% |   FAIL   |
|  I32  |     UNIQUE     |     0.7     |  75.820 ms |       0.09% |  75.761 ms |       0.03% |  -58.881 us |  -0.08% |   FAIL   |
|  I32  |     UNIQUE     |     0.8     |  77.103 ms |       0.04% |  77.069 ms |       0.04% |  -34.506 us |  -0.04% |   FAIL   |
|  I32  |     UNIQUE     |     0.9     |  81.284 ms |       0.02% |  81.283 ms |       0.03% |   -0.882 us |  -0.00% |   PASS   |
|  I64  |     UNIQUE     |     0.1     |  81.933 ms |       1.95% |  81.870 ms |       1.90% |  -62.649 us |  -0.08% |   PASS   |
|  I64  |     UNIQUE     |     0.2     |  82.099 ms |       1.40% |  81.554 ms |       1.43% | -544.771 us |  -0.66% |   PASS   |
|  I64  |     UNIQUE     |     0.3     |  79.828 ms |       2.48% |  79.794 ms |       2.37% |  -34.489 us |  -0.04% |   PASS   |
|  I64  |     UNIQUE     |     0.4     |  77.414 ms |       0.07% |  77.352 ms |       0.07% |  -62.195 us |  -0.08% |   FAIL   |
|  I64  |     UNIQUE     |     0.5     |  77.081 ms |       0.06% |  77.089 ms |       0.07% |    8.197 us |   0.01% |   PASS   |
|  I64  |     UNIQUE     |     0.6     |  77.175 ms |       0.06% |  77.141 ms |       0.05% |  -33.949 us |  -0.04% |   PASS   |
|  I64  |     UNIQUE     |     0.7     |  77.549 ms |       0.04% |  77.485 ms |       0.04% |  -64.312 us |  -0.08% |   FAIL   |
|  I64  |     UNIQUE     |     0.8     |  78.887 ms |       0.03% |  78.805 ms |       0.04% |  -82.122 us |  -0.10% |   FAIL   |
|  I64  |     UNIQUE     |     0.9     |  83.173 ms |       0.04% |  83.124 ms |       0.05% |  -48.390 us |  -0.06% |   FAIL   |

# static_set_insert_gaussian_skew

## [0] Quadro RTX 8000

|  Key  |  Distribution  |  Skew  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|-------|----------------|--------|------------|-------------|------------|-------------|------------|---------|----------|
|  I32  |    GAUSSIAN    |  0.1   |  46.261 ms |       0.04% |  46.289 ms |       0.24% |  27.548 us |   0.06% |   FAIL   |
|  I32  |    GAUSSIAN    |  0.2   |  58.826 ms |       0.05% |  58.785 ms |       0.05% | -41.231 us |  -0.07% |   FAIL   |
|  I32  |    GAUSSIAN    |  0.3   |  64.349 ms |       0.12% |  64.304 ms |       0.02% | -44.795 us |  -0.07% |   FAIL   |
|  I32  |    GAUSSIAN    |  0.4   |  62.381 ms |       0.04% |  62.774 ms |       0.10% | 392.472 us |   0.63% |   FAIL   |
|  I32  |    GAUSSIAN    |  0.5   |  60.617 ms |       0.04% |  61.224 ms |       0.05% | 607.418 us |   1.00% |   FAIL   |
|  I32  |    GAUSSIAN    |  0.6   |  58.379 ms |       0.05% |  59.813 ms |       0.03% |   1.434 ms |   2.46% |   FAIL   |
|  I32  |    GAUSSIAN    |  0.7   |  57.828 ms |       0.04% |  59.196 ms |       0.05% |   1.368 ms |   2.37% |   FAIL   |
|  I32  |    GAUSSIAN    |  0.8   |  57.424 ms |       0.05% |  58.612 ms |       0.04% |   1.188 ms |   2.07% |   FAIL   |
|  I32  |    GAUSSIAN    |  0.9   |  56.677 ms |       0.08% |  57.966 ms |       0.04% |   1.289 ms |   2.27% |   FAIL   |
|  I32  |    GAUSSIAN    |   1    |  55.841 ms |       0.02% |  57.227 ms |       0.06% |   1.386 ms |   2.48% |   FAIL   |
|  I64  |    GAUSSIAN    |  0.1   |  48.207 ms |       0.10% |  49.179 ms |       6.43% | 972.640 us |   2.02% |   FAIL   |
|  I64  |    GAUSSIAN    |  0.2   |  60.591 ms |       0.09% |  60.559 ms |       0.18% | -31.971 us |  -0.05% |   PASS   |
|  I64  |    GAUSSIAN    |  0.3   |  66.242 ms |       0.08% |  66.205 ms |       0.08% | -36.417 us |  -0.05% |   PASS   |
|  I64  |    GAUSSIAN    |  0.4   |  64.132 ms |       0.06% |  64.548 ms |       0.08% | 416.271 us |   0.65% |   FAIL   |
|  I64  |    GAUSSIAN    |  0.5   |  62.333 ms |       0.11% |  62.955 ms |       0.06% | 622.068 us |   1.00% |   FAIL   |
|  I64  |    GAUSSIAN    |  0.6   |  60.131 ms |       0.06% |  61.586 ms |       0.06% |   1.455 ms |   2.42% |   FAIL   |
|  I64  |    GAUSSIAN    |  0.7   |  59.609 ms |       0.08% |  60.964 ms |       0.12% |   1.355 ms |   2.27% |   FAIL   |
|  I64  |    GAUSSIAN    |  0.8   |  59.178 ms |       0.10% |  60.444 ms |       0.13% |   1.266 ms |   2.14% |   FAIL   |
|  I64  |    GAUSSIAN    |  0.9   |  58.509 ms |       0.09% |  59.773 ms |       0.09% |   1.264 ms |   2.16% |   FAIL   |
|  I64  |    GAUSSIAN    |   1    |  58.226 ms |       2.78% |  59.051 ms |       0.08% | 824.666 us |   1.42% |   FAIL   |

Reply:
Any compiler worth it's salt should hoist that construction out of the loop